CLW/CLWParallelPrimitives.cpp

```diff
@@ -921,65 +920,181 @@ void CLWParallelPrimitives::ReclaimTempBuffer(std::map<size_t, CLWBuffer<T> > co
+    val.s[2] = std::numeric_limits<float>::min();
+
+    return val;
+}
+
+template <class T>
+CLWEvent CLWParallelPrimitives::Reduction(const char* kernelName,
+    unsigned int deviceIdx,
+    CLWBuffer<T> input,
+    int numElems,
+    CLWBuffer<T> out)
 {
-    assert(groupSize);
     assert(input.GetElementCount() >= numElems);
-    assert(output.GetElementCount() >= numElems);
 
-    int ELEMS_PER_WI = 4;
-    int GROUP_BLOCK_SIZE = (groupSize * ELEMS_PER_WI);
-    int NUM_BLOCKS = (numElems + GROUP_BLOCK_SIZE - 1) / GROUP_BLOCK_SIZE;
+    int NUM_BLOCKS = (int)((numElems + WG_SIZE - 1) / WG_SIZE);
 
-    CLWKernel normalizeKernel = program_.GetKernel(kernelName);
+    CLWKernel reductionKernel = program_.GetKernel(kernelName);
 
     int argc = 0;
 
-    normalizeKernel.SetArg(argc++, input);
-    normalizeKernel.SetArg(argc++, output);
-    normalizeKernel.SetArg(argc++, numElems);
-    normalizeKernel.SetArg(argc++, SharedMemory(2 * sizeof(T) * groupSize));
-    normalizeKernel.SetArg(argc++, cache);
+    reductionKernel.SetArg(argc++, input);
+    reductionKernel.SetArg(argc++, numElems);
+    reductionKernel.SetArg(argc++, SharedMemory(sizeof(T) * WG_SIZE));
+    reductionKernel.SetArg(argc++, out);
 
-    return context_.Launch1D(deviceIdx, NUM_BLOCKS * groupSize, groupSize, normalizeKernel);
+    return context_.Launch1D(deviceIdx, NUM_BLOCKS * WG_SIZE, WG_SIZE, reductionKernel);
 }
 
-CLWEvent CLWParallelPrimitives::Copy(unsigned int deviceIdx, CLWBuffer<cl_int> input, CLWBuffer<cl_int> output, int numElems)
+template <class T>
+CLWEvent CLWParallelPrimitives::Normalize(const char* normalizeKernelName,
+                                          const char* minReductionKernelName,
+                                          const char* maxReductionKernelName,
+                                          unsigned int deviceIdx,
+                                          CLWBuffer<T> input,
+                                          CLWBuffer<T> output,
+                                          int numElems,
+                                          CLWBuffer<T> cache)
 {
-    int ELEMS_PER_WI = 4;
-    int GROUP_BLOCK_SIZE = (WG_SIZE * ELEMS_PER_WI);
-    int NUM_BLOCKS = (numElems + GROUP_BLOCK_SIZE - 1) / GROUP_BLOCK_SIZE;
+    assert(input.GetElementCount() >= numElems);
+    assert(output.GetElementCount() >= numElems);
 
-    CLWKernel copyKernel = program_.GetKernel("copy");
+    int NUM_BLOCKS = (int)((numElems + WG_SIZE - 1) / WG_SIZE);
 
-    copyKernel.SetArg(0, input);
-    copyKernel.SetArg(1, numElems);
-    copyKernel.SetArg(2, output);
+    T min = GetMaxNum<T>();
+    T max = GetMinNum<T>();
 
-    return context_.Launch1D(0, NUM_BLOCKS * WG_SIZE, WG_SIZE, copyKernel);
+    context_.WriteBuffer<T>(deviceIdx, cache, &min, 1);
+
+   Reduction(minReductionKernelName,
+             0,
+             input,
+             numElems,
+             cache).Wait();
+
+    context_.ReadBuffer<T>(deviceIdx, cache, &min, 1).Wait();
```

Comment:
You should keep this data in GPUbuffer and do not read them back to CPU.

Reply:
Done